### PR TITLE
Rename Note Type: randomizeName if it is already exist

### PR DESCRIPTION
**Pull request by Akshay0701**
## Purpose / Description
In ankidroid whenever we rename with pre-existed note type then it gets saved with duplicate names 
whereas in AnkiDesktop it appends randomize keys to renamed string.

### CODE REVIEW

**Comment by Arthur-Milchior** [comment]
Would you mind splitting in two commits. One which create the function and move code in it, the other that do the actual functional change. It will be easier to check

File: AnkiDroid/src/main/java/com/ichi2/anki/ModelBrowser.java [hunk 1]
```diff
@@ -452,6 +471,11 @@ private void renameModelDialog() {
                                     String deckName = mModelNameInput.getText().toString()
                                             // Anki desktop doesn't allow double quote characters in deck names
                                             .replaceAll("[\"\\n\\r]", "");
+
+                                    if (mExistingModelsNames.contains(deckName)) {
+                                        deckName = randomizeName(deckName);
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/ModelBrowser.java:476]: `randomizeName` is likely incorrect. I would expect that the following should be ported to Models.java and used instead

https://github.com/ankidroid/Anki-Android/blob/ee7cd786ca991f23f8df17879816e80712fefe31/AnkiDroid/src/main/java/com/ichi2/libanki/ModelsV16.kt#L271-L281

File: AnkiDroid/src/main/java/com/ichi2/anki/ModelBrowser.java [hunk 2]
```diff
@@ -73,6 +73,8 @@
     private ArrayList<Integer> mCardCounts;
     private ArrayList<Long> mModelIds;
     private ArrayList<DisplayPair> mModelDisplayList;
+    private ArrayList<String> mNewModelLabels;
+    private ArrayList<String> mExistingModelsNames;
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/ModelBrowser.java:77]: mExistingModelNames

File: AnkiDroid/src/main/java/com/ichi2/anki/ModelBrowser.java [hunk 3]
```diff
@@ -400,6 +379,41 @@ private void addNewNoteType(String modelName, int position) {
...
+
+        //Used to fetch model names
+        mNewModelNames = new ArrayList<>(stdModelSize);
+        for (StdModels StdModels: StdModels.STD_MODELS) {
+            String defaultName = StdModels.getDefaultName();
+            mNewModelLabels.add(String.format(add, defaultName));
+            mNewModelNames.add(defaultName);
+        }
+
+        if (mModels != null) {
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/ModelBrowser.java:404]: Can `mModels` be null here? It doesn't feel possible

File: AnkiDroid/src/main/java/com/ichi2/anki/ModelBrowser.java [hunk 4]
```diff
@@ -302,36 +305,12 @@ private void updateSubtitleText() {
...
-
-        if (mModels != null) {
-            for (Model model : mModels) {
-                String name = model.getString("name");
-                newModelLabels.add(String.format(clone, name));
-                mNewModelNames.add(name);
-                existingModelsNames.add(name);
-            }
+        if (mExistingModelsNames == null || mNewModelLabels == null) {
+            return;
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/ModelBrowser.java:309]: When will this be re-called if the values are null?

**Review by david-allison-1 — COMMENTED** [review]
Could you split out the refactoring commit further. I don't understand how it works logically in its current form

File: AnkiDroid/src/main/java/com/ichi2/anki/ModelBrowser.java (hunk 1 above)
Comment (Akshay0701) [reply AnkiDroid/src/main/java/com/ichi2/anki/ModelBrowser.java:476]:
``` java
     /*
     * Takes current timestamp from col
     * Used to append to the end of new note types to dissuade
     * User from reusing names (which are technically not unique however
     */
    private String randomizeName(String s) {
        return s + " " + checksum(String.valueOf(getCol().getTime().intTimeMS())).substring(0, 5);
    }
```

File: AnkiDroid/src/main/java/com/ichi2/anki/ModelBrowser.java (hunk 4 above)
Comment (Akshay0701) [reply AnkiDroid/src/main/java/com/ichi2/anki/ModelBrowser.java:309]: I am returning method call if values are null and not recalling it

Comment (Akshay0701) [reply AnkiDroid/src/main/java/com/ichi2/anki/ModelBrowser.java:309]: I think this value will never become null, I should remove this if statement.

**Comment by Akshay0701** [comment]
> I don't understand how it works logically in its current form and it feels like a functional change

My mistake,

now I arranged commits in a more structured way,

first I fixed the randomizeName method (as David suggested),

then I extracted note type initialization code to a method
(because we going to need to call that in the `renameModelDialog` function).

and lastly, I added if statement in the renameModelDialog method to check if renamed string already exists or not.


let me know if there is any confusion left....

**Review by david-allison-1 — APPROVED** [review]
LGTM.

A unit test here would be great, as we're likely going to modify this functionality for the V16 conversion

**Review by mikehardy — APPROVED** [review]
Also LGTM - going to merge w/o test 🙈

**Comment by mikehardy** [comment]
This one took a conflict from #9286 / #9343 I believe - I think I have fixed them and this can squash merge if it passes CI

**Comment by mikehardy** [comment]
Hi there! **September OpenCollective Notice**

Apologies for the delay, and if you're already submitted something to OpenCollective for PRs merged in **September** 2021, you may feel free to ignore this and it will be processed shortly

Just a friendly notice that we try to process OpenCollective payments monthly - it's time for September 2021 submissions

If you are interested in compensation for this work, the process with details is here:

https://github.com/ankidroid/Anki-Android/wiki/OpenCollective-Payment-Process#how-to-get-paid

(I only post one comment per person to avoid spamming you, regardless of the number of PRs merged, but this note applies to all PRs merged for in the month of August)

Thanks!

**Pull request merged**